# Preprocessing 

```{r, echo=FALSE, results="hide"}
knitr::opts_chunk$set(error=FALSE, message=FALSE, warning=FALSE)
knitr::opts_chunk$set(dpi=300, dev="png", dev.args=list(pointsize=15))
library(BiocStyle)
```

## Loading in the data

The aim here is to convert the 10X data set into a `HDF5Matrix` object.
We first obtain the data using the `r Githubpkg("LTLA/TENxBrainData")` package:

```{r}
library(TENxBrainData)
sce <- TENxBrainData()
sce
```

We have a look at some of the cell-level metadata.
Data were obtained from multiple mice, which were captured and sequenced in multiple libraries. 
Note that the libraries and nested within the mice.

```{r}
table(sce$Library, sce$Mouse)
```

We also add some gene-level annotation.
There's already gene symbols, so we just add the chromosome location.

```{r}
library(TxDb.Mmusculus.UCSC.mm10.ensGene)
chr.loc <- mapIds(TxDb.Mmusculus.UCSC.mm10.ensGene, keys=rowData(sce)$Ensembl, 
                  keytype="GENEID", column="CDSCHROM")
rowData(sce)$Chr <- chr.loc
head(rowData(sce))
```

## Performing cell-based quality control

We use `r Biocpkg("scater")` to calculate quality control summarized for each cell and gene.

```{r}
library(scater)
sce <- calculateQCMetrics(sce, feature_controls=list(Mito=which(rowData(sce)$Chr=="chrM")))
head(colnames(rowData(sce)))
```

We have a look at the three relevant metrics.

```{r preprocess-qchist, fig.height=6, fig.width=12}
par(mfrow=c(1,3))
hist(sce$log10_total_counts, xlab=expression("Log"[10]~"library size"), col="grey80")
hist(sce$log10_total_features, xlab=expression("Log"[10]~"number of genes expressed"), col="grey80")
hist(sce$pct_counts_Mito, xlab="Percentage of mitochondrial reads", col="grey80", breaks=100)
```

We use some of these metrics for quality control on the cells.
This is done within each batch to avoid discarding cells, e.g., if one batch was sequenced at lower depth.

```{r}
low.libsize <- isOutlier(sce$log10_total_counts, nmad=3, batch=sce$Library, type="lower")
low.ngenes <- isOutlier(sce$log10_total_features, nmad=3, batch=sce$Library, type="lower")
discard <- low.libsize | low.ngenes 
data.frame(LowLib=sum(low.libsize), LowGenes=sum(low.ngenes), Lost=sum(discard))
```

Low-quality cells are discarded from the object.

```{r}
sce <- sce[,!discard]
```

## Looking at gene-based metrics

For further diagnostics, we calculate the average count (adjusted for library size) and the number of cells expressing each gene.

```{r}
ave.count <- calcAverage(sce)
rowData(sce)$AverageCount <- ave.count
hist(log10(ave.count), xlab=expression("Log"[10]~"average count"), col="grey80")
n.cells <- nexprs(sce, byrow=TRUE)
rowData(sce)$NumCells <- n.cells
hist(log10(n.cells), xlab=expression("Log"[10]~"number of cells"), col="grey80")
```

We also inspect the top most-highly-expressed genes.

```{r}
cols <- c("Ensembl", "Symbol", "AverageCount", "NumCells")
top.genes <- rowData(sce)[order(-ave.count),cols]
head(as.data.frame(top.genes), 20)
```

<!--
Serializing the SCE object:

```{r}
saveRDS(sce, file="objects/sce.rds")  
```
-->
